### PR TITLE
Cleanup host veth side when a pod is gone

### PR DESCRIPTION
**Pull request by phoracek**
With this patch, we assign pod side of veth pair not to pod netns directly, byt to a named netns linking to the pod's netns. By doing that, removal of pod, removes also host side of veth pair.

All we need to do is to clean up dead named netns symlinks.

Please, review only the second commit.

### CODE REVIEW

**Comment by phoracek** [comment]
In CNI bridge plugin, they create the pair inside container netns and then move one side to host, that may be even better solution without a need for named netns.

File: pkg/network/bridge/plugin.go [hunk 1]
```diff
@@ -274,8 +284,32 @@ func (nbdp *NetworkBridgeDevicePlugin) attachPodToBridge(bridgeName, nicName str
 		return err
 	}
 
+	// We use named ad-hoc created named netns linking to actual pod netns instead of the pod
+	// netns directly. Doing this, we ensure that both sides of veth pairs will be cleaned up
+	// once the pod is gone.
+	netnsName := fmt.Sprintf("%s%d", namespacePrefix, rand.Intn(999999999))
+
+	// ensure that /var/run/netns directory will be initialized
+	exec.Command("ip", "netns", "add", "dummyNS").Run()
```
Comment (yuvalif) [inline pkg/network/bridge/plugin.go]: why not use the netns library?

**Comment by yuvalif** [comment]
this is also what @rmohr  suggest here: https://github.com/kubevirt/kubevirt/pull/1407/files/4601c807d248223522166f336d218fffebfc0d6c..202946747ee3fe810a3a3724fcabc19d01a8f6f8#diff-5b80ff518958e93de348b14803f57c34
seems like the best option for cleanup.

**Comment by phoracek** [comment]
@yuvalif @rmohr I changed it, so it creates the veth inside pod netns, but it seems like it did not help, `nic_` stays on host until the DP is removed.

**Comment by rmohr** [comment]
@phoracek are you creating int in the namespace of the Pod of the DP itself or in the target Pod? The later is the correct way.

**Comment by phoracek** [comment]
@rmohr I use container PID to enter the namespace, then being there, I create veth pair and move veth peer to original (DP) namespace.

https://github.com/kubevirt/kubernetes-device-plugins/pull/50/commits/59c78e3c70df1f07a05a7aad03e7f5994561ec50#diff-4f9899f5b818b28f11bb7bebaaf54390R280

File: pkg/network/bridge/plugin.go [hunk 2]
```diff
@@ -237,107 +236,138 @@ func (nbdp *NetworkBridgeDevicePlugin) attachPods() {
...
-	// get namespace of the pod
-	ns, err := netns.GetFromPid(containerPid)
+	success = true
+
+	return nil
+}
+
+func inNetnsDo(targetNs netns.NsHandle, callback func(originalNs netns.NsHandle) error) error {
+	// store current namespace
+	originalNs, err := netns.Get()
```
Comment (yuvalif) [inline pkg/network/bridge/plugin.go]: maybe you need to create a specific namespace to be used by the DS pod on the host side, and not just use whatever you get from netns.Get()?this is the only difference i see from the other examples.

Comment (phoracek) [reply pkg/network/bridge/plugin.go]: There are two other differences.

One is that in other solutions, one side of veth remains in a namespace bound to a temporary one-shot process (CNI plugin), we use long running daemon. That would explain why we remove it only when the DS goes down.

Another difference might be that in some examples, they use named namespaces (in /var/run/netns/).

Comment (rmohr) [reply pkg/network/bridge/plugin.go]: So `virt-handler` is also long-running and it works in this PR https://github.com/kubevirt/kubevirt/pull/1407 flawless. There should also no connection to a process. 

Very unlikely, but did you check if your Pods are stuck in "terminating"? Maybe the pods do not get cleaned up?

Comment (phoracek) [reply pkg/network/bridge/plugin.go]: @rmohr  It took them quite a long time to die, I will check, it sounds likely.

Comment (phoracek) [reply pkg/network/bridge/plugin.go]: They are quite long in "terminating" but they end up removed, still not cleaned up.

Comment (yuvalif) [reply pkg/network/bridge/plugin.go]: @phoracek can you try to create a namespace for the host side as well?

Comment (phoracek) [reply pkg/network/bridge/plugin.go]: I will. I just rewrote it using CNI libraries, still not cleaning up.

Comment (phoracek) [reply pkg/network/bridge/plugin.go]: Finally with CNI libraries used, locking and Go 1.10 it works. Tests are raceful, but when tested manually, it works.

**Comment by phoracek** [comment]
Tests are passing now, cleanup works. Will finish this after #47

**Review by yuvalif — APPROVED** [review]

**Pull request closed without merging**